### PR TITLE
Use run_id to sort files

### PR DESCRIPTION
**Pull request by dschwoerer**
Also provide better error in some cases.

### CODE REVIEW

**Review by johnomotani — COMMENTED** [review]
@dschwoerer what is the use case motivating this change? I think the feature is nice, but the case I'd think of is where the run directories are not named in the expected way (e.g. `run0/`, `run1/`, etc.). For that this PR doesn't help (yet!). It should be possible to get `run_restart_from` as well as `run_id` and then (assuming that the runs are a set of consecutive restarts, which presumably they usually will be if they are being concatenated in the time dimension) sort `run_ids`, `run_restart_from_list` and `paths_sorted` so that `run_restart_from_list[i] == run_ids[i-1]`.

Also it would be good to test the new functionality in `test_load`. Not sure how complicated that would be though, I guess you would need to add some functionality to generate some `run_id` values.

**Comment by dschwoerer** [comment]
> @dschwoerer what is the use case motivating this change?

I was certain I had replied already. Sorry for not doing so.

The motivation is to allow something like `BOUT.dmp.*.nc` with having old, stale dump files.

> I think the feature is nice, but the case I'd think of is where the run directories are not named in the expected way (e.g. `run0/`, `run1/`, etc.). For that this PR doesn't help (yet!). It should be possible to get `run_restart_from` as well as `run_id` and then (assuming that the runs are a set of consecutive restarts, which presumably they usually will be if they are being concatenated in the time dimension) sort `run_ids`, `run_restart_from_list` and `paths_sorted` so that `run_restart_from_list[i] == run_ids[i-1]`.

That would also be nice but it is out of scope for this PR


> Also it would be good to test the new functionality in `test_load`. Not sure how complicated that would be though, I guess you would need to add some functionality to generate some `run_id` values.

I didn't get the time to figure out how the tests work. And the tests being horribly slow does not help :-(

**Comment by johnomotani** [comment]
> The motivation is to allow something like `BOUT.dmp.*.nc` with having old, stale dump files.

I'm not sure I'm quite following. This is already a nice improvement - that currently gives much nicer error messages, and could maybe be expanded to add more features - but it looks like this still opens all the files that are matched by the glob pattern. If I understand your comment, you want to handle something like running a simulation on 8 procs (the old stale simulation), then running a new one on 4 procs (the simulation we now want to open), and open with just `BOUT.dmp.*.nc` as the pattern. I agree this would be nice, but it seems like in this example e.g. `BOUT.dmp.7.nc` would still be added to `paths_grid` to be opened, although that is a 'stale file' - am I missing something?

File: xbout/load.py [hunk 1]
```diff
@@ -900,21 +900,79 @@ def _arrange_for_concatenation(filepaths, nxpe=1, nype=1):
...
+                )
+            paths = iter(paths)
 
-    # Create list of lists of filepaths, so that xarray knows how they should
-    # be concatenated by xarray.open_mfdataset()
-    paths = iter(filepaths)
-    paths_grid = [
-        [[next(paths) for x in range(nxpe)] for y in range(nype)] for t in range(n_runs)
-    ]
+            paths_grid.append([[next(paths) for x in range(nxpe)] for y in range(nype)])
```
Comment (dschwoerer) [inline xbout/load.py:975]: @johnomotani sorry for the slow reply.

No, the files are only added unconditionally if the runs do not contain runids.

If they do, the above code filters out stale files.

**Review by johnomotani — APPROVED** [review]
I see now, thanks @dschwoerer!
Looks good to me.

**Comment by dschwoerer** [comment]
I would have liked to add tests, but I doubt to get to that anytime soon :-(

**Pull request merged**